create the workflow and add parsing for GitHub content+ Contribution frequency

PaintedW0lf opened the pull request:
<!-- 
Thank you for contributing! Please fill out this template to help us review your PR.
-->

## 📝 Description

> Please include a summary of the change and which issue is fixed.  
> Include relevant motivation and context.  
> List any dependencies that are required for this change.

This PR implements individual work parsing for non-code files by distinguishing between collaborative and non-collaborative documentation, then applying different parsing strategies based on authorship.
it's now a  Fully Dynamic system to handle different ways for parsing each type of file : 
✅ Works with git repos → extracts commit counts and author contributions
✅ Works without git → uses fallback values, skips collaborative parsing
✅ Handles errors gracefully → catches exceptions, returns defaults
✅ No manual configuration needed → automatically detects git availability

Last-minute changes added #351 for the changed output for the parser to change the output to this for the changes requirements

Expected output from parsing
{
  "path": "/test/project_proposal.pdf",
  "name": "project_proposal.pdf",
  "type": "pdf",
  "content": "..USER CONTRIBUTION. extracted text ...",
  "success": true,
  "error": "",
  "contribution_frequency” :  number
}
and updated all the test cases as well 


### Solution Overview

**1. File Classification (`non_code_file_checker.py`)**
- `get_classified_non_code_file_paths()` now returns a dictionary with separate lists:
  - `collaborative`: Files with 2+ authors where user is one of them
  - `non_collaborative`: Files with only the user as author OR local files
- `verify_user_in_files()` updated to place README files in `non_collaborative` list to resolve the bug

**2. Differential Parsing Strategy (`document_parser.py`)**
- `parsed_input_text()` now accepts a dictionary instead of a flat list
- **Non-collaborative files**: Extracts FULL content from the actual file
- **Collaborative files**: Extracts ONLY author's git diffs/patches
- **README files**: Always get full content parsing (excluded from git diff extraction)

**3. Git Diff Extraction (`git_utils.py` + `document_parser.py`)**
- `extract_non_code_content_by_author()` filters non-code files by author
- `parse_author_contributions()` explicitly excludes README files (`exclude_readme=True`)
- Returns git patches showing only what the author added/modified

**4. Integration (`main.py`)**
- Updated to handle dictionary return format from classification
- Passes classified files to parser with repo path and author email
- Maintains backward compatibility with analysis pipeline


### CONTRIBUTION FREQUENCY LOGIC - 
The contribution frequency tracks how many times a user touched/modified a specific file across all their commits. Here's the implementation:

In document_parser.py - parse_author_contributions() function:
Extracts git commits for the author using extract_non_code_content_by_author() from git_utils
Counts commits per file using a dictionary

### Key Points:
Each commit where the user modified a file = +1 to frequency

If a user modified report.md in 5 different commits → contribution_frequency: 5

For non-collaborative files (solo work), frequency is always 1 (full ownership)

For collaborative files, it shows actual git history touch count



**Closes:** #344 , #342, #351, #352, #365- bug 
#367 for test updates

---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [ ] ♻️ Refactoring
- [ ] ⚡ Performance improvement

---

## 🧪 Testing

> Please describe how you tested this PR (both manually and with tests).  
> Provide instructions so we can reproduce.


**Manual Testing:**
- [x] Tested git metadata extraction for author contributions
- [x] Verified author email filtering from user preferences
- [x] Confirmed integration with non-code analysis pipeline
- [x] Validated parsing of individual vs collaborative content

**Automated Testing:**
- [x] Unit tests for git author extraction functions
- [x] Unit tests for `extract_non_code_content_by_author()`
- [x] Integration tests with analysis pipeline

- `test_individual_work_parsing.py` - Basic classification and parsing
- `test_collaborative_parsing.py` - Git diff extraction verification
- `test_readme_correct.py` - README full content verification - [#365](https://github.com/COSC-499-W2025/capstone-project-team-3/issues/365), this needs for changing readme to non colaborative 
- `test_final_verification.py` - Comprehensive end-to-end test


**To reproduce:**

docker compose exec server pytest tests/ -v -k "non_code"



## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
-[x] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [x] 🔗 Any dependent changes have been merged and published in downstream modules
- [ ] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile



## 📸 Screenshots

> If applicable, add screenshots to help explain your changes

<details>
<summary>Click to expand screenshots</summary>

<!-- Add your screenshots here -->

</details>

dabby04 commented on app/utils/non_code_parsing/document_parser.py:
```diff
@@ -18,9 +18,10 @@
     docx = None
 
-def parse_documents_to_json(file_paths, output_path):
+def parse_documents_to_json(file_paths, output_path, repo_path=None, author=None):
```
We don't necessarily have repo_path at all, this would be checked when we detect git on a file path. Could we possibly remove  `repo_path` as a parameter since we never really know the path until we get_repo?

PaintedW0lf replied:
u r right, this should happen later in case our flow doesnt have git non git seperation earlier

dabby04 reviewed (changes requested):
Hi @PaintedW0lf, thank you for your contribution. Some of the changes you made to document_parser are causing the previous tests you had called `test_non_codeParser.py` to fail. 

Asides from this, I had a few questions:
- Why might we need to exclude files when extracting non code content by author
- I don't understand the purpose of is_author_only, wouldn't all author's contributions be extracted before then?

6s-1 reviewed (commented):
As discussed offline, as suggested by you we might need to pass the number of times the user touched the file even if they are the only one who contributed to it, please implement what you think is correct.

abstractafua commented on app/utils/non_code_analysis/non_code_analysis_utils.py:
```diff
@@ -422,6 +424,10 @@ def create_non_code_analysis_prompt(aggregated_project_metrics):
 
     # Append file-level details to the prompt
     PROMPT += "\nFile Details:\n" + "\n".join(file_details)
+    
```
Prompt modification shouldn't be needed at this stage we should only be sending content that has been made by the individual.

PaintedW0lf commented:
@dabby04 
Thanks so much for ur comments. I fixed the tests now , as the output needed for analysis was changed i went back to change the previous tests now. Please let me know if there is any other questions. 
1) We exclude README and PDF/DOCX files because:
README files - Already handled separately by extract_all_readmes() in git_utils.py. we will parse the whole content here to present the good summary 
PDF/DOCX files - Binary files that git can't show meaningful diffs for. When you commit a PDF, git just shows "binary file changed" - no actual content diff. So extracting "author contributions" from git for these files is useless.
2) removed

PaintedW0lf commented:
Thanks @dabby04 for noticing, all fixed now! Hope the logic makes sense

PaintedW0lf commented on app/utils/git_utils.py:
```diff
@@ -454,6 +454,112 @@ def extract_pull_request_metrics(
+                status = "A" if d.new_file else "D" if d.deleted_file else "R" if d.renamed_file else "M"
+                patch_text = getattr(d, "diff", b"")
+                try:
+                    patch = patch_text.decode("utf-8", errors="replace")
```
Good question! The default `exclude_pdf_docx=True` is intentional because:

1. **Git diffs don't work for binary files**: PDFs and DOCX are binary formats. Trying to decode their git diffs as UTF-8 text produces garbled output.

2. **Different parsing strategy**: PDFs/DOCX are parsed separately through `document_parser.py` using specialized libraries (`pypdf`, `python-docx`) to extract actual text content, not through git diffs.

3. **This function is for git diff extraction only**: `extract_non_code_content_by_author()` extracts git patches for text-based files (`.md`, `.txt`). For binary files, we use the full content extraction pathway in `document_parser.py`.

So the architecture is:
- Text files (`.md`, `.txt`): Git diffs work → use this function
- Binary files (`.pdf`, `.docx`): Git diffs don't work because we want to parse the whole content we consider it solo → use `document_parser.py` with specialized parsers

The default keeps the function focused on what it does well (text diffs) and avoids returning unusable binary data.

i would be happy to get on a call to describe the whole logic as needed

PaintedW0lf commented on app/utils/non_code_analysis/non_code_file_checker.py:
```diff
@@ -246,20 +247,26 @@ def verify_user_in_files(
+    return {
+        "collaborative": classification["collaborative"],
+        "non_collaborative": classification["non_collaborative"]
+    }
```
@kjassani Based on offline comment - here is what is sent! 
@6s-1 also worked in the same file to seperate out excluded files. We can all discuss here!
